tiledb/array.py

```diff
@@ -243,3 +251,748 @@ def index_domain_subarray(array, dom, idx: tuple):
+        """
+        Consolidates fragments of an array object for increased read performance.
+
+        Overview: https://docs.tiledb.com/main/concepts/internal-mechanics/consolidation
+
+        :param tiledb.Config config: The TileDB Config with consolidation parameters set
+        :param key: (default None) encryption key to decrypt an encrypted array
+        :type key: str or bytes
+        :param fragment_uris: (default None) Consolidate the array using a list of fragment _names_ (note: the `__ts1_ts2_<label>_<ver>` fragment name form alone, not the full path(s))
+        :param timestamp: (default None) If not None, consolidate the array using the given tuple(int, int) UNIX seconds range (inclusive). This argument will be ignored if `fragment_uris` is passed.
+        :type timestamp: tuple (int, int)
+        :raises: :py:exc:`tiledb.TileDBError`
+
+        Rather than passing the timestamp into this function, it may be set with
+        the config parameters `"sm.vacuum.timestamp_start"`and
+        `"sm.vacuum.timestamp_end"` which takes in a time in UNIX seconds. If both
+        are set then this function's `timestamp` argument will be used.
+
+        """
+        if self.mode == "r":
+            raise tiledb.TileDBError(
+                "cannot consolidate array opened in readonly mode (mode='r')"
+            )
+
+        if config is None:
+            config = tiledb.Config(self.ctx.config())
+
+        if key is not None:
+            config["sm.encryption_type"] = "AES_256_GCM"
+            config["sm.encryption_key"] = key
+
+        if fragment_uris is not None:
+            if timestamp is not None:
+                warnings.warn(
+                    "The `timestamp` argument will be ignored and only fragments "
+                    "passed to `fragment_uris` will be consolidate",
+                    DeprecationWarning,
+                )
+            self.array._consolidate(self.ctx, fragment_uris, config)
+            return
+        elif timestamp is not None:
+            warnings.warn(
+                "The `timestamp` argument is deprecated; pass a list of "
+                "fragment URIs to consolidate with `fragment_uris`",
+                DeprecationWarning,
+            )
+
+            if (
+                not isinstance(timestamp, tuple)
+                and len(timestamp) != 2
+                and not (
+                    isinstance(timestamp[0], int) and isinstance(timestamp[1], int)
+                )
+            ):
+                raise TypeError(
+                    "'timestamp' argument expects tuple(start: int, end: int)"
+                )
+
+            if timestamp[0] is not None:
+                config["sm.consolidation.timestamp_start"] = timestamp[0]
+            if timestamp[1] is not None:
+                config["sm.consolidation.timestamp_end"] = timestamp[1]
+
+        self.array._consolidate(self.ctx, config)
+
+    def upgrade_version(self, config=None):
```

Comment:
In the future this should be updated to accept a `self_or_uri` argument, like we did for `delete_fragments`. Same with `consolidate` above, but that might need to be deprecated because we already have `tiledb.consolidate`.

Reply:
Story opened!